### PR TITLE
Initialization

### PR DESCRIPTION
**Pull request by varisd**
Making initialization great again

### CODE REVIEW

File: neuralmonkey/decoders/decoder.py [hunk 1]
```diff
@@ -257,26 +256,26 @@ def embedding_matrix(self) -> tf.Variable:
...
     @tensor
     def decoding_b(self) -> tf.Variable:
         with tf.name_scope("output_projection"):
             return tf.get_variable(
-                "state_to_word_b", [len(self.vocabulary)],
-                initializer=tf.constant_initializer(
-                    - math.log(len(self.vocabulary))))
+                name="state_to_word_b",
+                shape=[len(self.vocabulary)],
+                initializer=tf.zeros_initializer())
```
Comment (jlibovicky) [inline neuralmonkey/decoders/decoder.py]: Tady bych to nerušil.

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: máš odkaz na něco, co to obhajuje? Experimenty s nulama běží úplně stejně jako s logem vokabuláře.

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: protože mi to přijde jako šarlatánství a všichni ostatní lidi světa dávaj do všech biasů nuly kromě některých uvnitř LSTM sítí, kdy inicializace biasů jedničkou zajistí nějaký defaultní chování těch gatů.

Comment (jlibovicky) [reply neuralmonkey/decoders/decoder.py]: Ondra kdysi říkal, že to doporučuje Otrhan Firat, ale nedaří se mi to rychle vygooglit. Ale přijde mi, že tohle je jediná inicializace, která naopak není šarlatánství. Je to jediný parametr, pro který existuje jasná interpretace, co to dělá: je to apriorní pravděpodobnost slov na výstupu, a když pro ní máme jednoduchý odhad, tam mi přijde škoda toho nevyužít.

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: Tady je argumentace proč tohle vyhodit:

softmax(x + b) = (e^xi * e^b) / (sum (e^xj * e^b)) = (e^b/e^b) * (e^xi / sum (e^xj)) = softmax(x)

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: (https://plus.google.com/+AlbertZeyer/posts/N2wG8xuCAKZ)

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: Je to __nenormalizovaná__ apriorní pravděpodobnost. Pokud by to měla bejt pravděpodobnost, necpal bys to s tim zbytkem do softmaxu, ne?

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: pokud bys chtěl použít apriorní pravděpodobnosti, vynásobil bys je s výsledkem softmaxu, nevynásobil bys?

Comment (jlibovicky) [reply neuralmonkey/decoders/decoder.py]: To nemůžeš vytknou, protože to b je pro každý slovo ze slovníku jiný.

Comment (jlibovicky) [reply neuralmonkey/decoders/decoder.py]: Ta diskuze, co odkazuješ mluví o tom, jak by to vypadalo, kdybysme to inicializovali log/(1 / velikost slovníku)

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: No na začátku pro všechny stejný je. Derivace budou vycházet furt stejně, protože derivace logitu vůči biasu je jedna.

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: (softmax asi bude distributivní, ale jistej si nejsem - tim by neplatil muj kec začínající "pokud bys")

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: Firat kterej udělal nematus tam má nuly.

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: Argument o tom, že tyhle parametry jsou ty nejjednodušší na naučení mi taky přijde validní.

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: ad diskuse - já vim, proto jsem jí zmínil, protože mu tam veskrze píšou, že je to jedno.

Comment (jlibovicky) [reply neuralmonkey/decoders/decoder.py]: Ale tohle je vektor, kde je v každym políčku logaritmus pravděpodobnosti toho slova, takže to nejsou stejný čísla.

Comment (varisd) [reply neuralmonkey/decoders/decoder.py]: Souhlasím s tím, že zero_initializer by měl stačit. To vlákno, které Jindra postnul výše, to dobře obhajuje.

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: Ještě jedna matika. Pokud začnu s biasama nastavenýma na jednu stejnou konstantu pro všechny slova _b_, pak softmax(x+b) = softmax(x). Pak dostanu update vektor _u_ takovej, že b1 = b + u. Potom softmax(x + b1) = softmax(x + b + u) = softmax(x + u), protože b je vektor stejných čísel. Víme, že gradient v biasech nezáleží na biasech, takže z toho plyne, že je to jedno.

Comment (jlibovicky) [reply neuralmonkey/decoders/decoder.py]: To je to, o čem celou dobu mluvím.

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: V tom případě ti nerozumim.

Comment (jlibovicky) [reply neuralmonkey/decoders/decoder.py]: Ale to vlákno mluví o tom, když je ten vektor nastavenej na jednu stejnou konstantu!

Comment (jlibovicky) [reply neuralmonkey/decoders/decoder.py]: A jestli konstantu nebo nuly je jedno, ale jestli distribuci nebo konstantu není jedno.

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: `initializer=tf.constant_initializer( BLABLABLA )` je jedna stejná konstanta

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: logaritmus délky slovníku je přece pro všechny slova, sovy, králíky i princezny stejnej, ne?

Comment (jlibovicky) [reply neuralmonkey/decoders/decoder.py]: Pardón, jsem se překouk, myslel jsem, že tam jsou pravděpodobnosti.

**Comment by tomkocmi** [comment]
This was bad decision. As for the embeddings, I did a lot of experiments last year (https://arxiv.org/abs/1711.09160) and the best initialization was to use as small as possible, even zeros worked really good. Therefore moving the initialization of embeddings from 0.001 to 0.01 is not a good idea. Also the Glorot aka Xavier initialization have some prerequisites where it is the best to use IN THEORY. The main condition is that it suppose linear layer. The best initialization for ReLu layers is He et al. (2015). There are other initializations for CNN, for RNN but basically no one found the Holy Grail of initialization since for each type works something different.
Therefore making all initialization the same was not a good idea. Even worse if you decided to put 0.01 there, which forces NN to make a lot augmentation in comparison to 0.001. 
Do you have any reasons, why was this changed? Any experiments that showed this setup has better or the same results in comparison to previous one? It could be the reason why my models based on config from last year have 6 BLEU instead of 16 BLEU (ps: I already tried the different output projections as discussed at lunch and it did not helped)
